README.md
id-jclient
==========

Java client for [https://github.com/btnguyen2k/id-server](https://github.com/btnguyen2k/id-server).

## License ##

See [LICENSE.txt](LICENSE.txt) for details. Copyright (c) 2015 btnguyen2k.

Third party libraries are distributed under their own license(s).

## Release-notes ##

Latest release: `0.1.0`.

See [RELEASE-NOTES.md](RELEASE-NOTES.md).


RELEASE-NOTES.md
id-jclient Release Notes
========================

Java client for [https://github.com/btnguyen2k/id-server](https://github.com/btnguyen2k/id-server).

2015-05-05: v0.1.0
------------------
First release:

- REST & Thrift client


src/test/java/qnd/QndRestIdClient.java
package qnd;

import com.github.btnguyen2k.id.jclient.IIdClient;
import com.github.btnguyen2k.id.jclient.impl.RestIdClient;
import com.github.btnguyen2k.id.jclient.impl.RestIdClientFactory;

public class QndRestIdClient {

    public static void main(String[] args) {
        IIdClient idClient = new RestIdClient().setIdServerUrl("http://localhost:9000").init();
        System.out.println(idClient.nextId("default"));

        idClient = RestIdClientFactory.newIdClient("http://localhost:9000");
        System.out.println(idClient.nextId("default"));
    }

}


src/test/java/qnd/QndThriftIdClient.java
package qnd;

import com.github.btnguyen2k.id.jclient.IIdClient;
import com.github.btnguyen2k.id.jclient.impl.ThriftIdClient;
import com.github.btnguyen2k.id.jclient.impl.ThriftIdClientFactory;

public class QndThriftIdClient {

    public static void main(String[] args) {
        IIdClient idClient = new ThriftIdClient().setIdServerHost("localhost")
                .setIdServerPort(9090).init();
        System.out.println(idClient.nextId("default"));

        idClient = ThriftIdClientFactory.newIdClient("localhost:9090");
        System.out.println(idClient.nextId("default"));
    }

}


src/test/java/qnd/BaseQndThriftClient.java
package qnd;

public class BaseQndThriftClient {
}


src/main/java/com/github/btnguyen2k/id/jclient/impl/RestIdClient.java
package com.github.btnguyen2k.id.jclient.impl;

import java.util.Map;

import jodd.http.HttpRequest;
import jodd.http.HttpResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.github.btnguyen2k.id.jclient.EngineType;
import com.github.btnguyen2k.id.jclient.IIdClient;
import com.github.ddth.commons.utils.DPathUtils;
import com.github.ddth.commons.utils.SerializationUtils;

/**
 * REST-implementation of {@link IIdClient}.
 * 
 * @author Thanh Nguyen <btnguyen2k@gmail.com>
 * @since 0.1.0
 */
public class RestIdClient extends AbstractIdClient {

    private Logger LOGGER = LoggerFactory.getLogger(RestIdClient.class);

    private String idServerUrl;

    public RestIdClient() {
    }

    public RestIdClient(String idServerUrl) {
        setIdServerUrl(idServerUrl);
    }

    public String getIdServerUrl() {
        return idServerUrl;
    }

    public RestIdClient setIdServerUrl(String idServerUrl) {
        this.idServerUrl = idServerUrl;
        if (this.idServerUrl.endsWith("/")) {
            this.idServerUrl = this.idServerUrl.substring(0, this.idServerUrl.length() - 1);
        }
        return this;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> callApi(String url) {
        try {
            HttpResponse httpResponse = HttpRequest.get(url).timeout(5000).send();
            try {
                if (httpResponse.statusCode() != 200) {
                    return null;
                }

                int contentLength = Integer.parseInt(httpResponse.contentLength());
                if (contentLength == 0 || contentLength > 1024) {
                    LOGGER.warn("Invalid response length: " + contentLength);
                    return null;
                }

                return SerializationUtils.fromJsonString(httpResponse.body(), Map.class);
            } finally {
                httpResponse.close();
            }
        } catch (Exception e) {
            LOGGER.warn(e.getMessage(), e);
            return null;
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public long nextId(String namespace) {
        return nextId(namespace, EngineType.DEFAULT);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public long nextId(String namespace, EngineType engine) {
        String apiUri = "/nextId/" + namespace;
        if (engine != null) {
            switch (engine) {
            case JDBC:
                apiUri = "/nextId/" + namespace + "/jdbc";
                break;
            case REDIS:
                apiUri = "/nextId/" + namespace + "/redis";
                break;
            case SNOWFLAKE:
                apiUri = "/nextId/" + namespace + "/snowflake";
                break;
            case ZOOKEEPER:
                apiUri = "/nextId/" + namespace + "/zookeeper";
                break;
            default:
            }
        }
        Map<String, Object> apiResult = callApi(idServerUrl + apiUri);
        if (apiResult == null) {
            return -1;
        }
        Integer status = DPathUtils.getValue(apiResult, "status", Integer.class);
        Long id = DPathUtils.getValue(apiResult, "id", Long.class);
        return status != null && status.intValue() == 200 ? (id != null ? id.longValue() : -1) : -1;
    }
}


src/main/java/com/github/ddth/id/jclient/thrift/ThriftClientFactory.java
package com.github.ddth.id.jclient.thrift;

import org.apache.thrift.protocol.TCompactProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TFramedTransport;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

import com.github.btnguyen2k.idserver.thrift.TIdService;
import com.github.ddth.thriftpool.ITProtocolFactory;
import com.github.ddth.thriftpool.PoolConfig;
import com.github.ddth.thriftpool.ThriftClientPool;

/**
 * Factory to create id-server Thrift clients.
 * 
 * @author ThanhNB
 * @since 0.1.0
 */
public class ThriftClientFactory {

    /**
     * Helper method to create a new {@link ITProtocolFactory} for id-server
     * Thrift client.
     * 
     * @param host
     * @param port
     * @return
     */
    public static ITProtocolFactory protocolFactory(final String host, final int port) {
        ITProtocolFactory protocolFactory = new ITProtocolFactory() {
            @Override
            public TProtocol create() {
                TSocket socket = new TSocket(host, port);
                socket.setTimeout(10000);
                TTransport transport = new TFramedTransport(socket);
                TProtocol protocol = new TCompactProtocol(transport);
                return protocol;
            }
        };
        return protocolFactory;
    }

    /**
     * Helper method to create a new client-pool with default settings.
     * 
     * @param host
     * @param port
     * @return
     */
    public static ThriftClientPool<TIdService.Client, TIdService.Iface> newDefaultClientPool(
            final String host, final int port) {
        final ThriftClientPool<TIdService.Client, TIdService.Iface> pool = new ThriftClientPool<TIdService.Client, TIdService.Iface>();
        pool.setClientClass(TIdService.Client.class).setClientInterface(TIdService.Iface.class);
        pool.setTProtocolFactory(protocolFactory(host, port));
        pool.setPoolConfig(new PoolConfig().setMaxActive(32).setMaxWaitTime(10000));
        pool.init();
        return pool;
    }
}


src/main/java/com/github/btnguyen2k/id/jclient/impl/RestIdClientFactory.java
package com.github.btnguyen2k.id.jclient.impl;

import java.util.concurrent.ExecutionException;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.cache.RemovalListener;
import com.google.common.cache.RemovalNotification;

/**
 * Factory to create {@link RestIdClient}.
 * 
 * @author Thanh Nguyen <btnguyen2k@gmail.com>
 * @since 0.1.0
 */
public class RestIdClientFactory {
    private static LoadingCache<String, RestIdClient> cache = CacheBuilder.newBuilder()
            .removalListener(new RemovalListener<String, RestIdClient>() {
                @Override
                public void onRemoval(RemovalNotification<String, RestIdClient> notification) {
                    notification.getValue().destroy();
                }
            }).build(new CacheLoader<String, RestIdClient>() {
                @Override
                public RestIdClient load(String idServerUrl) throws Exception {
                    RestIdClient idClient = new RestIdClient();
                    idClient.setIdServerUrl(idServerUrl).init();
                    return idClient;
                }
            });

    public static void cleanup() {
        cache.invalidateAll();
    }

    /**
     * Helper method to create a new {@link RestIdClient} instance.
     * 
     * @param idServerUrl
     * @return
     */
    public static RestIdClient newIdClient(String idServerUrl) {
        try {
            return cache.get(idServerUrl);
        } catch (ExecutionException e) {
            return null;
        }
    }
}


src/main/java/com/github/btnguyen2k/id/jclient/impl/ThriftIdClientFactory.java
package com.github.btnguyen2k.id.jclient.impl;

import java.util.concurrent.ExecutionException;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.cache.RemovalListener;
import com.google.common.cache.RemovalNotification;

/**
 * Factory to create {@link ThriftIdClient}.
 * 
 * @author Thanh Nguyen <btnguyen2k@gmail.com>
 * @since 0.1.0
 */
public class ThriftIdClientFactory {
    private static LoadingCache<String, ThriftIdClient> cache = CacheBuilder.newBuilder()
            .removalListener(new RemovalListener<String, ThriftIdClient>() {
                @Override
                public void onRemoval(RemovalNotification<String, ThriftIdClient> notification) {
                    notification.getValue().destroy();
                }
            }).build(new CacheLoader<String, ThriftIdClient>() {
                @Override
                public ThriftIdClient load(String hostAndPort) throws Exception {
                    ThriftIdClient idClient = new ThriftIdClient(hostAndPort);
                    idClient.init();
                    return idClient;
                }
            });

    public static void cleanup() {
        cache.invalidateAll();
    }

    /**
     * Helper method to create a new {@link ThriftIdClient} instance.
     * 
     * @param hostAndPort
     *            Thrift host & port of id-server, format {@code host:port}
     * @return
     */
    public static ThriftIdClient newIdClient(String hostAndPort) {
        try {
            return cache.get(hostAndPort);
        } catch (ExecutionException e) {
            return null;
        }
    }

    /**
     * Helper method to create a new {@link ThriftIdClient} instance.
     * 
     * @param idServerHost
     *            Thrift host of id-server
     * @param idServerPort
     *            Thrift port of id-server
     * @return
     */
    public static ThriftIdClient newIdClient(String idServerHost, int idServerPort) {
        return newIdClient(idServerHost + ":" + idServerPort);
    }
}


src/main/java/com/github/btnguyen2k/id/jclient/impl/ThriftIdClient.java
package com.github.btnguyen2k.id.jclient.impl;

import org.apache.commons.lang3.StringUtils;
import org.apache.thrift.protocol.TCompactProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TFramedTransport;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.github.btnguyen2k.id.jclient.EngineType;
import com.github.btnguyen2k.id.jclient.IIdClient;
import com.github.btnguyen2k.idserver.thrift.TIdResponse;
import com.github.btnguyen2k.idserver.thrift.TIdService;
import com.github.ddth.thriftpool.ITProtocolFactory;
import com.github.ddth.thriftpool.PoolConfig;
import com.github.ddth.thriftpool.ThriftClientPool;

/**
 * Thrift-implementation of {@link IIdClient}.
 * 
 * @author Thanh Nguyen <btnguyen2k@gmail.com>
 * @since 0.1.0
 */
public class ThriftIdClient extends AbstractIdClient {

    private Logger LOGGER = LoggerFactory.getLogger(ThriftIdClient.class);

    private String idServerHost = "localhost";
    private int idServerPort = 9090;
    private ThriftClientPool<TIdService.Client, TIdService.Iface> thriftClientPool;

    /**
     * Constructs a new {@link ThriftIdClient} object.
     */
    public ThriftIdClient() {
    }

    /**
     * Constructs a new {@link ThriftIdClient} object.
     * 
     * @param idServerHost
     *            Thrift host of id-server
     * @param idServerPort
     *            Thrift port of id-server
     */
    public ThriftIdClient(String idServerHost, int idServerPort) {
        setIdServerHost(idServerHost);
        setIdServerPort(idServerPort);
    }

    /**
     * Constructs a new {@link ThriftIdClient} object.
     * 
     * @param hostAndPort
     *            Thrift host & port of id-server, format {@code host:port}
     */
    public ThriftIdClient(String hostAndPort) {
        String[] tokens = StringUtils.split(hostAndPort, ':');
        setIdServerHost(tokens[0]);
        if (tokens.length > 1) {
            setIdServerPort(Integer.parseInt(tokens[1]));
        }
    }

    public String getIdServerHost() {
        return idServerHost;
    }

    public ThriftIdClient setIdServerHost(String idServerHost) {
        this.idServerHost = idServerHost;
        return this;
    }

    public int getIdServerPort() {
        return idServerPort;
    }

    public ThriftIdClient setIdServerPort(int idServerPort) {
        this.idServerPort = idServerPort;
        return this;
    }

    /**
     * Helper method to create a new {@link ITProtocolFactory} for id-server
     * Thrift client.
     * 
     * @param host
     * @param port
     * @return
     */
    public static ITProtocolFactory protocolFactory(final String host, final int port,
            final int soTimeout) {
        ITProtocolFactory protocolFactory = new ITProtocolFactory() {
            @Override
            public TProtocol create() {
                TSocket socket = new TSocket(host, port);
                socket.setTimeout(soTimeout);
                TTransport transport = new TFramedTransport(socket);
                TProtocol protocol = new TCompactProtocol(transport);
                return protocol;
            }
        };
        return protocolFactory;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ThriftIdClient init() {
        super.init();

        final int timeout = 10000;
        thriftClientPool = new ThriftClientPool<TIdService.Client, TIdService.Iface>();
        thriftClientPool.setClientClass(TIdService.Client.class).setClientInterface(
                TIdService.Iface.class);
        thriftClientPool.setTProtocolFactory(protocolFactory(idServerHost, idServerPort, timeout));
        thriftClientPool.setPoolConfig(new PoolConfig().setMaxActive(32).setMaxWaitTime(timeout));
        thriftClientPool.init();

        return this;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void destroy() {
        if (thriftClientPool != null) {
            try {
                thriftClientPool.destroy();
            } catch (Exception e) {
            }
        }

        super.destroy();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public long nextId(String namespace) {
        return nextId(namespace, EngineType.DEFAULT);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public long nextId(String namespace, EngineType engine) {
        long id = -1;
        try {
            TIdService.Iface idClient = thriftClientPool.borrowObject();
            if (idClient != null) {
                try {
                    String _engine = "default";
                    if (engine != null) {
                        switch (engine) {
                        case JDBC:
                            _engine = "jdbc";
                            break;
                        case REDIS:
                            _engine = "redis";
                            break;
                        case SNOWFLAKE:
                            _engine = "snowflake";
                            break;
                        case ZOOKEEPER:
                            _engine = "zookeeper";
                            break;
                        default:
                        }
                    }
                    TIdResponse response = idClient.nextId(namespace, _engine);
                    id = response != null && response.status == 200 ? response.id : -1;
                } finally {
                    thriftClientPool.returnObject(idClient);
                }
            }
        } catch (Exception e) {
            LOGGER.warn(e.getMessage(), e);
        }
        return id;
    }
}
